# Workshop in computational analysis of ancient genome

## General info

**Date:** 15-17 November 2019

**Place:** Middle East Technical University, Ankara, Turkey

**Organized by:** [Comparative & Evolutionary Biology @ METU Biology](https://compevo.bio.metu.edu.tr/)

**Price:** Free

**Contact and application:** ancientdna.workshop@gmail.com

- All applicant must fill out the [application form](https://docs.google.com/forms/d/e/1FAIpQLSfNDIliERI2rWa_VFyJ8Di96vk9d3XfXke2p7nK6Fz6FDzjPg/viewform?vc=0&c=0&w=1) no later than November 10, 2019
- Excellence of motivation letter and relevance of applicants goals, study subject and past experience to topics covered in the workshop.
- Familiarity with statistical and computing languages
- Past attendance to such courses/seminars/workshops (first time applicants will be given priority)
- Institutional and cultural diversity
- Priority will be given to graduate students (MSc/PhD) and early stage post-docs.
- You should bring a **laptop** to the course.

**Instructors**
- Dilek Koptekin, Department of Health Informatics, Graduate School of Informatics, METU
- Ekin Sağlıcan, Department of Biological Sciences, METU
- Kıvılcım Başak Vural, Department of Biological Sciences, METU
- Mehmet Somel, Department of Biological Sciences, METU

## Course material
The lecture will be based on a large amount of reading material that should be read in advance - you can find them below. The slides used during the lectures will be made available right before the lectures.

### Pre-requisites:
- [Unix tutorial](https://people.ischool.berkeley.edu/~kevin/unix-tutorial/toc.html)
- [Command-line tutorial](https://rik.smith-unna.com/command_line_bootcamp/?id=pabgahcnti)
- Learn to keep notes! (use efficient text editor i.e. [atom](https://atom.io/))

### Softwares and their publications:

* **bwa:** [Page/Manual](http://bio-bwa.sourceforge.net/bwa.shtml), [Publication(1)](https://www.ncbi.nlm.nih.gov/pubmed/19451168), [Publication(2)](https://www.ncbi.nlm.nih.gov/pubmed/20080505)
* **samtools:** [Page/Manual](http://www.htslib.org/doc/samtools.html), [Publication(1)](https://www.ncbi.nlm.nih.gov/pubmed/19505943), [Publication(2)](https://www.ncbi.nlm.nih.gov/pubmed/21903627)
* **plink:** [Page/Manual](http://zzz.bwh.harvard.edu/plink/pdf.shtml), [Publication(1)](https://www.ncbi.nlm.nih.gov/pubmed/17701901)
* **vcftools:** [Page/Manual](https://vcftools.github.io/documentation.html), [Publication(1)](https://academic.oup.com/bioinformatics/article/27/15/2156/402296)
* **PMDtools:** [Page/Manual](https://code.google.com/archive/p/pmdtools/), [Publication(1)](https://www.ncbi.nlm.nih.gov/pubmed/?term=Separating+endogenous+ancient+DNA+from+modern+day+contamination+in+a+Siberian+Neandertal)
* **contaMix:** [Publication(1)](https://www.ncbi.nlm.nih.gov/pubmed/?term=revised+timescale+for+human+evolution+based+on+ancient+mitochondrial+genomes)
* **schmutzi:** [Page/Manual](https://bioinf.eva.mpg.de/schmutzi/), [Publication(1)](https://www.ncbi.nlm.nih.gov/pubmed/26458810)
* **ANGSD:** [Page/Manual](https://github.com/ANGSD/angsd), [Publication(1)](https://www.ncbi.nlm.nih.gov/pubmed/25420514)
* **EIGENSOFT:** [Page/Manual](https://github.com/DReichLab/EIG), [Publication(1)](https://www.ncbi.nlm.nih.gov/pubmed/16862161), [Publication(2)](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC1713260/)
* **Admixtools:** [Page/Manual](https://github.com/DReichLab/AdmixTools), [Publication(1)](https://www.ncbi.nlm.nih.gov/pubmed/22960212)
* **ADMIXTURE:** [Page/Manual](http://software.genetics.ucla.edu/admixture/admixture-manual.pdf), [Publication(1)](https://www.ncbi.nlm.nih.gov/pubmed/?term=D.H.+Alexander%2C+J.+Novembre%2C+and+K.+Lange.+Fast+model-based+estimation+of+ancestry+in+unrelated+individuals)

### Data format
  - [fasta](https://www.animalgenome.org/bioinfo/resources/manuals/seqformats)
  - [fastq](https://pythonhosted.org/OBITools/fastq.html)
  - [bam/sam](https://sites.google.com/site/bioinformaticsremarks/bioinfo/sam-bam-format)
  - [bed](https://www.ensembl.org/info/website/upload/bed.html)
  - [vcf](https://samtools.github.io/hts-specs/VCFv4.2.pdf)
  - [plink (ped/map-tped/tfam-bed/bim/fam)](http://zzz.bwh.harvard.edu/plink/pdf.shtml)
  - [eigenstrat (geno/snp/ind)](https://reich.hms.harvard.edu/software/InputFileFormats)

### **MUST READ!** - Archaeogenomics studies on Anatolia

* [Mathieson et al 2015](https://www.nature.com/articles/nature16152)
* [Omrak et al 2016](https://www.cell.com/action/showPdf?pii=S0960-9822%2815%2901516-X)
* [Kılınç et al 2016](https://www.cell.com/action/showPdf?pii=S0960-9822%2816%2930850-8)
* [Lazaridis et al 2016](https://www.nature.com/articles/nature19310)
* [Hofmanová et al 2016](https://www.pnas.org/content/113/25/6886.long)
* [Kılınç et al 2017](https://royalsocietypublishing.org/doi/full/10.1098/rspb.2017.2064)
* [Damgaard et al 2018](https://science.sciencemag.org/content/360/6396/eaar7711)
* [Feldman et al 2019](https://www.nature.com/articles/s41467-019-09209-7)

## Quiz!
